### PR TITLE
[#48] morley optimizer integration

### LINKED ISSUE #48: Incorporate optimizer into our pipeline
# Clarification and motivation

[Morley](https://gitlab.com/morley-framework/morley) has an optimizer that can turn a contract into an equivalent one with smaller size and gas consumption. We should use it to make our contract smaller.

# Acceptance criteria

1. `Makefile` exposes some way to build stablecoin contract and apply `morley optimize` command to it.
2. Our tests use the optimized version of the contract.

### PR DESCRIPTION
**Pull request by rvem**
## Description

This PR adds an ability to optimize generated ligo contract via `morley-optimize`.
An optimized contract is used as a test resource in CI.

## Related issue(s)

Resolves #48 

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock
  - [x] I updated [changelog](../tree/master/ChangeLog.md) unless I am sure my changes are
        not essential.

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,10 @@
 # SPDX-FileCopyrightText: 2020 tqtezos
 # SPDX-License-Identifier: MIT
 
-.PHONY: all build-ligo build-haskell test test-dumb-term test-hide-successes nettest clean
+.PHONY: all build-ligo build-haskell optimize-ligo test test-dumb-term test-hide-successes nettest clean
+
+# Ligo executable
```
Comment (gromakovsky) [inline Makefile:6]: :jebaited:

File: Makefile [hunk 2]
```diff
@@ -16,6 +19,10 @@ build-ligo:
 build-haskell:
 	$(MAKE_HASKELL) build
 
+# Optimitez built ligo contract using morley optimizer
```
Comment (gromakovsky) [inline Makefile:22]: > Optimitez

Wut?

File: haskell/Makefile [hunk 3]
```diff
@@ -28,7 +29,7 @@ endef
 define build_stablecoin_contract
 	$(MAKE_LIGO) stablecoin.tz
 	mkdir -p test/resources
-	ln -sf "../../$(LIGO_DIR)/stablecoin.tz" "test/resources/stablecoin.tz"
+	morley optimize --contract "$(LIGO_DIR)/stablecoin.tz" --output "test/resources/stablecoin.tz"
```
Comment (gromakovsky) [inline haskell/Makefile:32]: You define `MORLEY` but use `morley`.

File: nix/sources.json [hunk 4]
```diff
@@ -40,5 +40,16 @@
...
         "url": "https://github.com/serokell/tezos-packaging/archive/8e7e651e5a02858ffbb9900fb200498777651df6.tar.gz",
         "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
+    },
+    "morley": {
+        "branch": "master",
+        "description": "Developer tools for the Michelson Language",
+        "homepage": "",
+        "owner": "morley-framework",
+        "repo": "morley",
+        "rev": "f61e31dfd4977350cbc1f3fb23392c8799bc2fe4",
```
Comment (gromakovsky) [inline nix/sources.json:50]: Erm, it's very sad that now `morley` revision appears in two places. Strictly speaking, it doesn't have to match the one from `stack.yaml`, so it's not that big of a problem, but still sad.
1. Can we somehow extract this information from `stack.yaml` instead?
2. If we can't, I find it better to use `pkgs.haskellPackages.morley`. Right now `morley` on Hackage has an issue that it depends on `tezos-bake-monitor-lib`, but we are very close to eliminating this dependency. Once we finish eliminating it, we will make a new release which will hopefully be usable via `pkgs.haskellPackages.morley`, at least that's how I understand it.

File: .buildkite/pipeline.yml [hunk 5]
```diff
@@ -35,7 +37,8 @@ steps:
    commands: &run-nettest
      - nix build -L -f. nettest
      - mkdir -p test/resources/
-     - nix-build -A tezos-contract -o test/resources/stablecoin.tz
+     - nix-build -A tezos-contract -o stablecoin.tz
+     - nix run -f. morley -c morley optimize --contract stablecoin.tz --output test/resources/stablecoin.tz
```
Comment (gromakovsky) [inline .buildkite/pipeline.yml]: Now I see some non-trivial amount of duplication between this step and the one where we run `stablecoin-test`. Is it possible to deduplicate common parts?

File: Makefile (hunk 2 above)
Comment (rvem) [reply Makefile:22]: I've tried :joy:

File: Makefile (hunk 1 above)
Comment (rvem) [reply Makefile:6]: :yao:

File: nix/sources.json (hunk 4 above)
Comment (rvem) [reply nix/sources.json:50]: > Can we somehow extract this information from stack.yaml instead?

Looks hacky, nix dependencies doesn't have anything to do with stack dependencies

**Comment by gromakovsky** [comment]
>  fixup! [#48] Add optimitez call to the Makefile 

You could have used `--squash` here btw. In that case during `git rebase --autosquash` you would be able to fix the commit message :)

File: .buildkite/pipeline.yml [hunk 6]
```diff
@@ -4,6 +4,8 @@
 
 env:
   NIX_PATH: nixpkgs=https://github.com/nixos/nixpkgs/archive/b640dbd00877f51616750e347a32fbcb49d2968c.tar.gz
+  FETCH_CONTRACT:
+    "mkdir -p test/resources/ && buildkite-agent artifact download stablecoin.tz test/resources/ --step \"LIGO contract\""
```
[the remark this answers is not in the document]
Comment (rvem) [reply .buildkite/pipeline.yml]: I think that it's third time, when I'm upset because of list concatenation absence

File: nix/sources.json (hunk 4 above)
Comment (gromakovsky) [reply nix/sources.json:50]: https://gitlab.com/morley-framework/morley/-/merge_requests/456

Comment (gromakovsky) [reply nix/sources.json:50]: Can you try `morley-1.4.0`?

**Comment by gromakovsky** [comment]
> error: Package ‘morley-0.6.0’ in /nix/store/y38lsywxy4dnpmnlmpgm1fndjcx87h5y-source/pkgs/development/haskell-modules/hackage-packages.nix:167947 is marked as broken, refusing to evaluate.

Why is it trying 0.6.0? That's fairly old.

**Comment by gromakovsky** [comment]
Also I suggest rebasing.

**Comment by rvem** [comment]
> Why is it trying 0.6.0?
That's because of haskell.nix adds haskell packages that are in lts-15.13. However, morley isn't there, so it's using very outdated morley that appeared in nixpkgs.

File: .buildkite/pipeline.yml [hunk 7]
```diff
@@ -35,8 +46,7 @@ steps:
      MONEYBAG: "unencrypted:edsk3GjD83F7oj2LrnRGYQer99Fj69U2QLyjWGiJ4UoBZNQwS38J4v"
    commands: &run-nettest
      - nix build -L -f. nettest
-     - mkdir -p test/resources/
-     - nix-build -A tezos-contract -o test/resources/stablecoin.tz
+     - bash -c "$FETCH_CONTRACT"
```
Comment (gromakovsky) [inline .buildkite/pipeline.yml:49]: Isn't it better to call `/usr/bin/env bash` here and above? It probably shouldn't affect anything, just a matter of style.

File: .buildkite/pipeline.yml [hunk 8]
```diff
@@ -20,11 +24,18 @@ steps:
  - label: build library
    if: *not_scheduled
    command: nix build -L -f. lib
+ - label: LIGO-contract
+   if: build.source != "schedule"
```
Comment (gromakovsky) [inline .buildkite/pipeline.yml:28]: Why not `   if: *not_scheduled`?

Comment (gromakovsky) [reply .buildkite/pipeline.yml:28]: Also are we sure it's not needed for scheduled builds?

Comment (rvem) [reply .buildkite/pipeline.yml:28]: hmm, this actually looks like a rebase artifact

File: .buildkite/pipeline.yml (hunk 7 above)
Comment (rvem) [reply .buildkite/pipeline.yml]: I think that `eval "$FETCH_CONTRACT"` better suits here, dunno, why I chose `bash -c` here

File: default.nix [hunk 9]
```diff
@@ -39,11 +42,15 @@ let
     hs-pkgs = project;
     local-packages = local-packages;
   };
+  morley =
+    (pkgs.haskell-nix.hackage-package
+      { name = "morley"; version = "1.4.0"; compiler-nix-name = "ghc883"; }
```
Comment (gromakovsky) [inline default.nix:47]: Btw, I wonder which optimization settings will be used when `morley` is built.

Comment (rvem) [reply default.nix:47]: AFAIU, cabal-to-nix uses `cabal v2-configure` which has `-O1` by default

**Pull request merged**